packages/arb-token-bridge-ui/src/token-bridge-sdk/OftTransferStarter.ts

```diff
@@ -0,0 +1,209 @@
+import { constants, ethers, Signer } from 'ethers'
+import { Provider } from '@ethersproject/providers'
+import { ERC20__factory } from '@arbitrum/sdk/dist/lib/abi/factories/ERC20__factory'
+import {
+  BridgeTransferStarter,
+  TransferEstimateGas,
+  TransferProps,
+  TransferType,
+  ApproveTokenProps,
+  RequiresTokenApprovalProps,
+  BridgeTransferStarterProps
+} from './BridgeTransferStarter'
+import { fetchErc20Allowance } from '../util/TokenUtils'
+import { getAddressFromSigner } from './utils'
+import { getOftTransferConfig, buildSendParams, getOftQuote } from './oftUtils'
+import oftAbi from './oftAbi.json'
+import { isNetwork } from '../util/networks'
+
+export class OftTransferStarter extends BridgeTransferStarter {
+  public transferType: TransferType = 'oft'
+  private isOftTokenValidated?: boolean
+  private oftAdapterAddress?: string
+  private destLzEndpointId?: number
+  private isSourceChainEthereum?: boolean
+
+  constructor(props: BridgeTransferStarterProps) {
+    super(props)
+    if (!this.sourceChainErc20Address) {
+      throw Error('OFT token address not found')
+    }
+  }
+
+  private async validateOftTransfer() {
+    if (typeof this.isOftTokenValidated !== 'undefined') {
+      if (!this.isOftTokenValidated) {
+        throw Error('Token is not supported for OFT transfer')
+      }
+      return
+    }
+
+    if (!this.sourceChainErc20Address) {
+      this.isOftTokenValidated = false
+      throw Error('OFT token address not found')
+    }
+
+    const [sourceChainId, destinationChainId] = await Promise.all([
+      this.sourceChainProvider.getNetwork().then(n => n.chainId),
+      this.destinationChainProvider.getNetwork().then(n => n.chainId)
+    ])
+
+    const oftTransferConfig = getOftTransferConfig({
+      sourceChainId,
+      destinationChainId,
+      sourceChainErc20Address: this.sourceChainErc20Address
+    })
+
+    if (!oftTransferConfig.isValid) {
+      this.isOftTokenValidated = false
+      throw Error('Token is not supported for OFT transfer')
+    }
+
+    this.isSourceChainEthereum = !!isNetwork(sourceChainId).isEthereumMainnet
+    this.isOftTokenValidated = true
+    this.oftAdapterAddress = oftTransferConfig.sourceChainAdapterAddress
+    this.destLzEndpointId = oftTransferConfig.destinationChainLzEndpointId
+  }
+
+  private getContractAddress(): string {
+    if (!this.isOftTokenValidated) {
+      throw Error('OFT validation not performed')
+    }
+    return this.oftAdapterAddress!
+  }
+
+  private getContract(providerOrSigner: Signer | Provider): ethers.Contract {
+    return new ethers.Contract(
```

Comment:
every time we're instantiating a new one, which I don't think we need to. we can store it as a property the first time and afterwards return that